Document VerificationExceptions and add failure section to docs (#27478)
gradle/gradle: platforms/documentation/docs/src/docs/userguide/authoring-builds/tasks/custom_tasks.adoc

In order to support cancellation (e.g. when the user stops the build with CTRL+C) and task timeouts, custom tasks should react to their executing thread being interrupted.
The same is true for work items submitted via the worker API. If a task does not respond to an interrupt within 10s, the daemon will shut down in order to free up system resources.

== Verification Failures

> Exceptions thrown during task execution result in task failure which typically immediately terminates a build with a build failure result (except when running with the <<command_line_interface.adoc#sec:continue_build_on_failure, `--continue` flag>>).

Reviewer: ```suggestion
Exceptions thrown during task execution result in a failure which immediately terminates a build with a build failure result (except when running with the <<command_line_interface.adoc#sec:continue_build_on_failure, `--continue` flag>>).
```

Another exception to this rule is when a task throws a link:{javadocPath}/org/gradle/api/tasks/VerificationException.html--[VerificationException].
These exceptions are used to indicate that a task has failed in a controlled manner such that its output is still valid.
Consumer tasks that use this output can still run.

> A failed test task, for instance, will cause a failing outcome for the test task.  However, this shouldn't prevent another task from reading and processing the (valid) test results output files it produced.

Reviewer: ```suggestion
A failed test task, for instance, will cause a failing outcome for the test task.  
However, this shouldn't prevent another task from reading and processing the (valid) resulting output files the test produced.
```

In order to support cancellation (e.g. when the user stops the build with CTRL+C) and task timeouts, custom tasks should react to their executing thread being interrupted.
The same is true for work items submitted via the worker API. If a task does not respond to an interrupt within 10s, the daemon will shut down in order to free up system resources.

== Verification Failures

> Exceptions thrown during task execution result in a failure that immediately terminates a build.

Reviewer 2: ```suggestion
Normally, exceptions thrown during task execution result in a failure that immediately terminates a build.
```

> The result of the build will be `FAILED`, and no further tasks will be executed.

Reviewer 2: ```suggestion
The outcome of the task will be `FAILED`, the result of the build will be `FAILED`, and no further tasks will be executed.
```

> When <<command_line_interface.adoc#sec:continue_build_on_failure, running with the `--continue` flag>> Gradle will continue to run other requested tasks in the build after encountering a task failure.

Reviewer 2: ```suggestion
When <<command_line_interface.adoc#sec:continue_build_on_failure, running with the `--continue` flag>>, Gradle will continue to run other requested tasks in the build after encountering a task failure.
```

However, any tasks that depend on a failed task will be not be executed.

> There is a special type of exception that behaves differently.

Reviewer 2: ```suggestion
There is a special type of exception that behaves differently when downstream tasks only rely on the outputs of a failing task.
```

Reviewer 2: And maybe something like...
> A task depends on the _outcome_ of another task when it directly depends on it with `dependsOn`.

> A task can throw a subtype of link:{javadocPath}/org/gradle/api/tasks/VerificationException.html--[VerificationException] to indicate that it has failed in a controlled manner such that its output is still valid for consumers.

Reviewer: The `VerificationException` [link](https://builds.gradle.org/repository/download/Gradle_Release_Check_BuildDistributions/76138450:id/distributions/gradle-8.6-docs.zip!/gradle-8.6-20231220211144%2B0000/docs/javadoc/org/gradle/api/tasks/VerificationException.html--) didn't work for me in the [build](https://builds.gradle.org/repository/download/Gradle_Release_Check_BuildDistributions/76138450:id/distributions/gradle-8.6-docs.zip!/gradle-8.6-20231220211144%2B0000/docs/userguide/custom_tasks.html#verification_failures)

Author: Got it.

When Gradle is run with `--continue`, consumer tasks that depend on this output can still run.

A failed unit test, for instance, will cause a failing outcome for the test task.

> However, this doesn't prevent another task from reading and processing the (valid) test results output files the task produced.

Reviewer: This is the only sentence that is hard to read. Can we do something like this:
```suggestion
However, this doesn't prevent another task from reading and processing the (valid) test results the task produced.
```

Verification failures are used in exactly this manner by the <<test_report_aggregation_plugin.adoc#test_report_aggregation_plugin, `Test Report Aggregation Plugin`>>.

> Verification failures are also useful for tasks that produce output piece by piece that is consumable after any number of pieces.

Reviewer 2: I'm not sure what this is trying to say.  Without verification failures, you can still have multiple tasks that produce something piece-by-piece.  The only difference is if those tasks can fail and still produce something useful.
